version 4/03 environments/04 Production Configuration/README.md

```diff
@@ -0,0 +1,341 @@
+const merge = require('webpack-merge');
+const common = require('./base.webpack.config.js');
+
+module.exports = merge(common, {
+    mode: 'development',
+    devtool: 'inline-source-map',
+    devServer: {
+        contentBase: './dist'
+    }
+});
+```
+
+- Make a configuration to see in console the enviroment in which we are
+
+### ./averageService.js
+```diff
+export function getAvg(scores) {
+  return getTotalScore(scores) / scores.length;
+}
+
+export function getTotalScore(scores) {
+  return scores.reduce((score, count) => score + count);
+}
+
++ console.log(`We are in: ${process.env.NODE_ENV}`);
+```
+
+- Finally we need to update command script:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+-   "start": "webpack-dev-server --mode development --open --hot",
++    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+    "build": "rimraf dist && webpack --mode development"
+  },
+  ...
+}
+```
+
+- Running `npm start`:
+
+We can see in console `We are in: development`
+
+- We are going to create a `build dev` command to see how much size has bundles files without `webpack-dev-server` stuff:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+-    "build": "rimraf dist && webpack --mode development"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js"
+  },
+  },
+  ...
+}
+
+```
+
+- Running `npm run build`:
+
+The `dist` folder is created. If we execute `index.html`, it shows us `We are in: development`.
+
+- Let's configure `production` environment:
+
+### ./prod.webpack.config.js
+```javascript
+const merge = require('webpack-merge');
+const common = require('./base.webpack.config.js');
+
+module.exports = merge(common, {
+    mode: 'production',
+});
+
+```
+
+- Add production command script:
+
+### ./package.json
+```diff
+{
+  ...
+  "scripts": {
+    "start": "webpack-dev-server --open --config dev.webpack.config.js",
+-    "build": "rimraf dist && webpack --config dev.webpack.config.js"
++    "build": "rimraf dist && webpack --config dev.webpack.config.js",
++    "build:prod": "rimraf dist && webpack --config prod.webpack.config.js"
+  },
+  ...
+}
+
+```
+
+- Running `npm run build:prod`, we can see in console `We are in: production` and we can see how all bundle sizes decrease:
+
+`npm run build`
+```
+  app.js          =>    46 KB
+  app.css         =>     2 KB
+  index.html      =>     1 KB
+  vendor.js       =>  2060 KB
+  vendor.css      =>   462 KB
+  vendorStyles.js =>    15 KB
+
+```
+
+`npm run build:prod`
+```
+  0.css           => 170 KB
+  app.js          =>  12 KB
+  app.css         =>   1 KB
+  index.html      =>   1 KB
+  vendor.css      => 108 KB
+  vendorStyles.js =>   2 KB
+
+```
+
+- As an optional configuration, we can go one over step and generate gzipped bundles. We need install [compression-webpack-plugin](https://github.com/webpack-contrib/compression-webpack-plugin):
```

Comment:
Just as an optional extra step, we could generated the gzipped bundles (usually this is just handled by the web servers but is a good idea to generate this files in order to check the size). We need to install [compression-webpack-plugin](https://github.com/webpack-contrib/compression-webpack-plugin):